src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs

```diff
@@ -93,53 +97,125 @@ public sealed class CosmosFhirDataStore : IFhirDataStore, IProvideCapability
             EnsureArg.IsNotNull(resource, nameof(resource));
 
             var cosmosWrapper = new FhirCosmosResourceWrapper(resource);
+            var partitionKey = new PartitionKey(cosmosWrapper.PartitionKey);
 
-            try
-            {
-                _logger.LogDebug($"Upserting {resource.ResourceTypeName}/{resource.ResourceId}, ETag: \"{weakETag?.VersionId}\", AllowCreate: {allowCreate}, KeepHistory: {keepHistory}");
+            _logger.LogDebug($"Upserting {resource.ResourceTypeName}/{resource.ResourceId}, ETag: \"{weakETag?.VersionId}\", AllowCreate: {allowCreate}, KeepHistory: {keepHistory}");
 
-                UpsertWithHistoryModel response = await _retryExceptionPolicyFactory.GetRetryPolicy().ExecuteAsync(
-                    async ct => await _upsertWithHistoryProc.Execute(
-                        _containerScope.Value.Scripts,
-                        cosmosWrapper,
-                        weakETag?.VersionId,
-                        allowCreate,
-                        keepHistory,
-                        ct),
-                    cancellationToken);
-
-                return new UpsertOutcome(response.Wrapper, response.OutcomeType);
+            if (weakETag == null && allowCreate && !cosmosWrapper.IsDeleted)
+            {
+                // Optimistically try to create this as a new resource
+                try
+                {
+                    await _retryExceptionPolicyFactory.GetRetryPolicy().ExecuteAsync(
+                        async ct => await _containerScope.Value.CreateItemAsync(
+                            cosmosWrapper,
+                            partitionKey,
+                            cancellationToken: ct,
+                            requestOptions: new ItemRequestOptions { EnableContentResponseOnWrite = false }),
+                        cancellationToken);
+
+                    return new UpsertOutcome(cosmosWrapper, SaveOutcomeType.Created);
+                }
+                catch (CosmosException e) when (e.StatusCode == HttpStatusCode.Conflict)
+                {
+                    // this means there is already an existing version of this resource
+                }
             }
-            catch (CosmosException exception)
+
+            while (true)
```

Comment:
Could you change this to while (!cancellationToken.IsCancellationRequested), or add a counter to limit retries? If the Cosmos DB SDK had an update with a bug that stopped it from respecting cancellation tokens, or some other bug is introduced that would ignore them, we could have an infinite loop here. Those are never fun to chase down in prod :)

Reply:
Good point. Added a ThowIfCancellationRequested() in the beginning of the loop.